### PR TITLE
[v2] Migrate semantic (binder) snapshot tests

### PR DESCRIPTION
**Pull request by ggiraldez**
Builds on top of #1577 

Adds semantic snapshot tests for v2. These are a port of the old bindings_output reports, with the following changes:

- Remove obsolete snapshots targeting Solidity <0.8.0.
- Change IR visitor to visit identifier terminals, which is needed to properly generate the reports.
- Simplify IR mutator by removing `is_removed` attribute for fields, which is not needed.
- Allow IR mutator to insert new sequence fields in arbitrary order to preserve as much as possible visitor order wrt. the CST: in particular this impacts the order in which identifiers can be collected so the reports maintain the definition numeric designations from v1.

### CODE REVIEW

File: crates/codegen-v2/semantic/src/ir/mutator.rs [hunk 1]
```diff
@@ -362,11 +349,9 @@ impl IrModelMutator {
         assert!(removed, "Could not find type {name} to remove");
 
         for sequence in self.sequences.values_mut() {
-            for field in &mut sequence.fields {
-                if field.target_type == identifier {
-                    field.is_removed = true;
```
Comment (teofr) [inline crates/codegen-v2/semantic/src/ir/mutator.rs:367]: Is there any reason we needed `is_removed` before (in v1) but not anymore?

Comment (ggiraldez) [reply crates/codegen-v2/semantic/src/ir/mutator.rs:367]: In the builder from the v1 CST it's needed to skip over the removed field when traversing the tree with a `Cursor`.

It was also used for skipping over removed fields in the transformer, but I don't think that was really necessary.

File: crates/solidity-v2/outputs/cargo/tests/src/semantic/semantic_output/report.rs [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+    report: &mut String,
+    file_id: &str,
+    contents: &str,
+    all_definitions: &[CollectedDefinition],
+    all_references: &[CollectedReference],
+    unbound_identifiers: &[CollectedIdentifier],
+    definitions_by_id: &HashMap<NodeId, usize>,
+) -> Result<()> {
+    // ariadne works with character offsets, not byte offsets, so we need to
+    // convert ranges
```
Comment (teofr) [inline crates/solidity-v2/outputs/cargo/tests/src/semantic/semantic_output/report.rs]: Could you add a TODO pointing to https://github.com/NomicFoundation/slang/issues/1536
We should fix this once available.

File: crates/solidity-v2/outputs/cargo/tests/src/semantic/semantic_output/report_data.rs [hunk 3]
```diff
@@ -0,0 +1,508 @@
...
+    }
+
+    fn current_file_contents(&self) -> &String {
+        let Some((_, contents)) = self.current_file else {
+            unreachable!("file is not set for collecting identifiers");
+        };
+        contents
+    }
+
+    fn byte_offset_to_line_column(&self, byte_offset: usize) -> (usize, usize) {
```
Comment (teofr) [inline crates/solidity-v2/outputs/cargo/tests/src/semantic/semantic_output/report_data.rs:118]: I think this is something our API should provide, not for this PR, but if you agree maybe a TODO here or an issue would be good.

Comment (teofr) [reply crates/solidity-v2/outputs/cargo/tests/src/semantic/semantic_output/report_data.rs:118]: Maybe even worth looking at [ariadne's](https://docs.rs/ariadne/latest/ariadne/struct.Source.html#method.get_offset_line) since it seems to do a binary search over the lines.

File: crates/solidity-v2/outputs/cargo/tests/src/semantic/semantic_output/report_data.rs [hunk 4]
```diff
@@ -0,0 +1,508 @@
...
+            all_references.push(CollectedReference {
+                identifier: identifier.clone(),
+                resolution: reference.resolution.clone(),
+            });
+            bound = true;
+        }
+
+        if !bound {
+            unbound_identifiers.push(identifier);
+        }
```
Comment (teofr) [inline crates/solidity-v2/outputs/cargo/tests/src/semantic/semantic_output/report_data.rs:190]: QQ, each identifier is either a definition, a reference, or unbound, right?

File: crates/solidity-v2/outputs/cargo/tests/src/utils/path_resolver.rs [hunk 5]
```diff
@@ -0,0 +1,82 @@
+/// Resolve an import path relative to a source file.
```
Comment (teofr) [inline crates/solidity-v2/outputs/cargo/tests/src/utils/path_resolver.rs:1]: Any reason not to use [`Path`](https://doc.rust-lang.org/std/path/struct.Path.html) across this file? I feel we're reinventing the wheel and missing some corner cases (for example, can we run this on Windows?)

**Review by teofr — COMMENTED** [review]
Some questions, but looking good

File: crates/solidity-v2/outputs/cargo/tests/src/semantic/semantic_output/report_data.rs [hunk 6]
```diff
@@ -0,0 +1,508 @@
...
+    }
+
+    (all_definitions, all_references, unbound_identifiers)
+}
+
+// Display helpers
```
Comment (teofr) [inline crates/solidity-v2/outputs/cargo/tests/src/semantic/semantic_output/report_data.rs:196]: Not for this PR, but I think it would make sense to have display methods for our various constructs

File: crates/solidity-v2/outputs/cargo/tests/src/semantic/semantic_output/report.rs (hunk 2 above)
Comment (ggiraldez) [reply crates/solidity-v2/outputs/cargo/tests/src/semantic/semantic_output/report.rs]: Nice, will do. I'm also noticing the comment is slightly misplaced and can be confusing.

File: crates/solidity-v2/outputs/cargo/tests/src/semantic/semantic_output/report_data.rs (hunk 3 above)
Comment (ggiraldez) [reply crates/solidity-v2/outputs/cargo/tests/src/semantic/semantic_output/report_data.rs:118]: IIUC ariadne can do a binary search because it previously split the lines and knows the offset each one starts at. I agree that we should provide this in the API, and I'm expecting we should do something similar then.

File: crates/solidity-v2/outputs/cargo/tests/src/semantic/semantic_output/report_data.rs (hunk 4 above)
Comment (ggiraldez) [reply crates/solidity-v2/outputs/cargo/tests/src/semantic/semantic_output/report_data.rs:190]: And identifier can be a definition and/or a reference, or be unbound. The typical case where it can be both is an unaliased import deconstruction symbol:
```solidity
import {Foo} from "foo.sol";
//      ^^^ this is both a reference (to the symbol in foo.sol) and a definition
import {Foo as Bar} from "foo.sol";
//             ^^^ this the definition
//      ^^^ this is the reference
```

I don't think there are other cases, but I don't recall off the top of my head.

File: crates/solidity-v2/outputs/cargo/tests/src/semantic/semantic_output/report_data.rs (hunk 6 above)
Comment (ggiraldez) [reply crates/solidity-v2/outputs/cargo/tests/src/semantic/semantic_output/report_data.rs:196]: I'm not sure about this. Unless it's for internal use, the display of a construct is use-case specific.

File: crates/solidity-v2/outputs/cargo/tests/src/semantic/semantic_output/report_data.rs (hunk 4 above)
Comment (teofr) [reply crates/solidity-v2/outputs/cargo/tests/src/semantic/semantic_output/report_data.rs:190]: Oh, that's right. Thanks!

File: crates/solidity-v2/outputs/cargo/tests/src/utils/path_resolver.rs (hunk 5 above)
Comment (ggiraldez) [reply crates/solidity-v2/outputs/cargo/tests/src/utils/path_resolver.rs:1]: This is used to normalize virtual paths (ie. we don't care about the underlying platform) that don't exist in the filesystem. The equivalent in `Path` would be [`Path::canonicalize`](https://doc.rust-lang.org/std/path/struct.Path.html#method.canonicalize) but the paths must exist. There's now a [`Path::normalize_lexically`](https://doc.rust-lang.org/std/path/struct.Path.html#method.normalize_lexically) which seems to do what we want, but didn't exist when we wrote this code and it's still experimental.

Comment (teofr) [reply crates/solidity-v2/outputs/cargo/tests/src/utils/path_resolver.rs:1]: That makes sense, thanks!
